### ISSUE TITLE
VarBuilder::from_bytes?

### ISSUE THREAD

**Issue by boustrophedon** [huggingface/candle#2177]
Is there a reason why `VarBuilder::from_buffered_safetensors` takes a Vec<u8> rather than a &[u8]?

Also, it's not clear to me why `VarBuilder::from_mmaped_safetensors` is unsafe. I don't think there's anything the user of the function can do to make their usage of it any more or less unsafe. Its safety seems to be entirely determined by the implementation of the function, which should be safe, and so the function shouldn't be marked unsafe.

**Comment by LaurentMazare** [comment]
Re `from_mmaped_safetensors` it's unsafe because it uses an unsafe function as [documented here](https://github.com/huggingface/candle/blob/d9bc5ec15164ecb583dbb25653edd89e08e6cbd0/candle-nn/src/var_builder.rs#L474) and you can see details on the root cause for this [here](https://docs.rs/memmap2/latest/memmap2/struct.MmapOptions.html#safety): it mostly boils down to what happens if the file disappears etc. We prefer passing the information of this potential unsafe behavior to the library caller so that they are aware of it.

**Comment by boustrophedon** [comment]
Thanks, that makes sense re: modifying a mmaped file. TIL!

I'm going to try to see if it's possible to implement a SimpleBackend for VarBuilder that uses a &[u8]

**Comment by LaurentMazare** [comment]
Yeah I don't think there is a good reason for not supporting the `&[u8]` it's probably just some historical stuff. I've put up #2179 which should provide the borrowed version (under a separate name so as not to break compatibility but maybe it should be the default).

**Comment by boustrophedon** [comment]
Thanks for doing this! I put up #2180 that makes a couple minor fixes

**Comment by boustrophedon** [comment]
Incidentally, the use case for this is that I can now do the following:

---
build.rs
```
fn main() {
    let api = hf_hub::api::sync::Api::new().unwrap();

    let api = api.repo(hf_hub::Repo::new(
        "example/model".to_string(),
         hf_hub::RepoType::Model,
    ));

    let model_file = api.get("model.safetensors").unwrap();

    println!("cargo::rustc-env=MODEL_FILE={}", model_file.display());
}
```

---
main.rs
```
static MODEL_DATA: &[u8] = include_bytes!(env!("MODEL_FILE"))

fn main() {
    [...]
    let vb = VarBuilder::from_slice_safetensors(MODEL_DATA, DType::F32, &device)?;
}
```
---

That is, you can download models at build time and then statically include them in the resulting binary. The same process works for including e.g. a `tokenizer.json`: you would output the filename in build.rs, use `include_bytes!()`, and then pass the static buffer to `Tokenizer::from_bytes`.

You could have a "bundled-model" feature so that if your model is large you don't have to copy it on every incremental compilation.

**Comment by boustrophedon** [comment]
Closed by #2179 and #2180 

Thanks again!
